Client receives notification when an irrelevant InitTx is observed

abailly-iohk opened the pull request:
This PR starts implementing #1096, tackling first the "simplest" transaction, `InitTx`.  The design is not as advanced as what is proposed an in [ADR 26](https://hydra.family/head-protocol/adr/26), but it seems a good first step in that direction. 

* We separate 2 phases of "observation", 
   * one that only depends on on-chain elements, eg. the `NetworkId` and the transaction itself, plus static information about the scripts, to produce a `RawInitObservation` that looks like an `InitTx` but might or might not be relevant to the observing node, 
   * one that takes a `RawInitObservation` and produces an `InitObservation` if the data in the transaction matches the `ChainContext` which contains the parameters of the head
* The "raw" observations are considered a different type of `ChainEvent` and are sent to the `ChainCallback` that's passed to the `Handlers`, as any other type of events. These raw observations are also logged
* Instead of being handled directly by the `HeadLogic.update`, these events are handled by a `Heads.update` function which "decorates" the former, with the same signature, and lets other events pass through. The intention is that the `Heads` module will handle all kind of inputs that are not related to a single Head's lifecycle, and in the future wrap multiple heads run concurrently inside a single node. Interestingly, this `Heads` module could also handle the connectivity related messages sent by the `Network` layer thus removing the need for a dedicated communication channel between the `Heartbeat` network component and the `Server` to notify clients of those events.

fix #529 

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

abailly committed:
Write ETE test for missed observation of wrong cardano keys

Test is failing

ffakenz committed:
Minor changes to simplify e2e spec

ffakenz committed:
Add ServerOutput type for SomeHeadInitializing

ffakenz committed:
Trace SomeHeadObserved when tx observation fails

abailly committed:
Provide a signaling component to ship logs to clients

abailly committed:
Expect headId in client side observation

abailly committed:
Dispatch non-relevant InitTx observations to clients

abailly committed:
Fix formatting

For some reason, fourmolu decided to format single constraints with
braces on my environment, perhaps following some version upgrade, so I
have added a configuration option to enforce no braces.

abailly committed:
Remove code duplication in hydra cluster configuration

abailly committed:
Update CHANGELOG

abailly committed:
Add observed pubkey hashes to the client's notification

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -219,6 +225,13 @@ spec = around showLogsOnFailure $
 
+      it "alice inits a Head with incorrect keys preventing bob from observing InitTx" $ \tracer ->
+        failAfter 60 $
+          withClusterTempDir "three-full-life-cycle" $ \tmpDir -> do
```
Should: Make sure to use a unique template. I belief "three-full-life-cycle" is still unchanged from copy/paste

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -654,6 +659,49 @@ initAndClose tmpDir tracer clusterIx hydraScriptsTxId node@RunningNode{nodeSocke
         failAfter 5 $ waitForUTxO networkId nodeSocket u
 
+initWithWrongKeys :: FilePath -> Tracer IO EndToEndLog -> RunningNode -> TxId -> IO ()
+initWithWrongKeys tmpDir tracer node@RunningNode{nodeSocket} hydraScriptsTxId = do
```
Could: move this scenario to the `Scenarios.hs` module

Not needed, just more consistent (although annoying as repl does not reload the library)

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -654,6 +659,49 @@ initAndClose tmpDir tracer clusterIx hydraScriptsTxId node@RunningNode{nodeSocke
+          _ -> chainConfig
+
+  withConfiguredHydraCluster tracer tmpDir nodeSocket 1 cardanoKeys hydraKeys hydraScriptsTxId wrongCardanoKeyForBob contestationPeriod $ \nodes -> do
+    let [n1, n2, n3] = toList nodes
```
Could: simplify this end to end scenario to only have two parties

Then, this additional function with a callback would not be needed and two invocations of `withHydraNode ... $ \n1 ->` would suffice. One of them with a misconfigured key of course. Furthermore, this would not also rely on `{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}`.. as we are keen on not relying on turned off warnings.

Also, instead of generating another key and writing it to disk, we could just mix up carol with bob/alice and expect the same behavior.

A comment (or haddock) on the test's strategy is warranted in any case and would be appreciated.

ch1bo commented on fourmolu.yaml:
```diff
@@ -6,3 +6,4 @@ diff-friendly-import-export: true # 'false' uses Ormolu-style lists
 respectful: true # don't be too opinionated about newlines etc.
 haddock-style: single-line # '--' vs. '{-'
 newlines-between-decls: 1 # number of newlines between top-level declarations
+single-constraint-parens: never # whether or not to put braces around single constraints: https://fourmolu.github.io/config/single-constraint-parens/
```
I think this did not apply as intended? Just below this change I see multiple of 
![image](https://github.com/input-output-hk/hydra/assets/2621189/0b9f8d1f-c022-47bb-a61f-babac02e5974)

ch1bo replied:
Let's tackle the formatting things in a separate PR explicitly checking format indempotence to the CI (CC @locallycompact)

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -89,6 +89,7 @@ data ServerOutput tx
     -- 'SnapshotConfirmed' message is emitted) UTxO's present in the Hydra Head.
     Greetings {me :: Party, headStatus :: HeadStatus, snapshotUtxo :: Maybe (UTxOType tx), hydraNodeVersion :: String}
   | PostTxOnChainFailed {postChainTx :: PostChainTx tx, postTxError :: PostTxError tx}
+  | SomeHeadInitializing {headId :: HeadId, pubKeyHashes :: [Text] }
```
Must: update the `api.yaml` to include this case.

I would further suggest a different name to indicate that this head was ignored (and has not been adopted / resulted in a state machine change)
```suggestion
  | IgnoredHeadInitializing {headId :: HeadId, pubKeyHashes :: [Text] }
```

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -352,7 +362,7 @@ prepareTxToPost timeHandle wallet ctx cst@ChainStateAt{chainState} tx =
   -- See ADR21 for context
   calculateTxUpperBoundFromContestationPeriod currentTime = do
-    let effectiveDelay = min (toNominalDiffTime $ contestationPeriod ctx) maxGraceTime
+    let effectiveDelay = min (toNominalDiffTime $ ctx.contestationPeriod) maxGraceTime
```
:open_mouth: do you want to use this more in the future? If yes, I would welcome a discussion and maybe a [coding standard](https://github.com/input-output-hk/hydra/wiki/Coding-Standards)?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -141,7 +142,7 @@ spec = parallel $ do
           genericCoverTable [transition] $
-            isJust (observeSomeTx ctx st tx)
+            isRight (observeSomeTx ctx st tx)
               & counterexample "observeSomeTx returned Nothing"
```
Should: update counter example text

ch1bo commented on hydra-node/src/Hydra/Logging/Signaling.hs:
```diff
@@ -0,0 +1,51 @@
+withSignaling ::
+  (MonadSTM m) =>
+  Tracer m (HydraLog tx net) ->
+  (Tracer m (HydraLog tx net) -> TVar m (Maybe (Server tx m)) -> m a) ->
```
Providing the TVar here on the callback is a bit odd and exposes internals. Instead, I would recommend returning only the `installSignal` callback function itself.

i.e. in usage:


```hs
withSignaling tracer $ \tracer' installSignal ->
  withChain tracer' ... $ \chain ->
    withServer tracer' chain ... $ \server -> 
      installSignal server
```

Also maybe even calling it `installSignalInServer server` in our usage site to make it clear that it's tying the knot between this signaling mechanism and the server.

ch1bo commented on hydra-node/test/Hydra/Logging/SignalingSpec.hs:
```diff
@@ -0,0 +1,30 @@
+          traceWith tracer (DirectChain $ SomeHeadObserved someHeadId ["12345"])
+          readTVarIO sentSignals
+
+    signalled `shouldBe` [SomeHeadInitializing someHeadId ["3132333435"]]
```
I'm a bit surprised that this passes? Reading this, I would expect the two singleton arrays to be equal `["12345"]`?

ch1bo reviewed (changes requested):
I do like the separation of "identifying an init tx" and "checking the init tx is for us" logic in the Tx module.

I do not like the `Signaling` approach. It appears to be generic, but then it's not as it's very specific to mapping one specific `HydraLog` into a `sendOutput` invocation of a `Server`.

Overall, this is a acceptable workaround for the mean-time to get some quick feedback when init transactions are ignored. The only "must" change requested from my side is to keep the `api.yaml` consistent by adding the new server output.

I will approve, _as long as_ we do agree that this is worsening the situation of callback spaghetti (providing the `chain` to `withServer` is another twist in this pasta dish).

Specifically, the way forward in my opinion is at odds with part of this PRs motivation:

> Using the HeadLogic to pass information to the client that's irrelevant to the state machine it implements would clutter the transitions and the code

IMO having a single flow of information and where decisions are made is beneficial. In fact, we are not only passing information but actually take the decision of "not adopting" this head, but ignoring it (which ultimately produces a server output).

---

Nitpick: fourmolu was not applied to `Main.hs` and maybe other modules. (I think I should open a PR to have this as a CI step instead of commenting here)

abailly committed:
Move wrong keys scenario to Scenarios file

abailly committed:
Simplify 'incorrect keys' scenario to only use 2 nodes

abailly committed:
Fix formatting

abailly committed:
Rename and document new output for ignored head

abailly committed:
Remove language extension

abailly committed:
Fix counterexample

abailly committed:
Remove Signaling in favor of a dedicated event to be handled by node

abailly committed:
Layer update function to handle non-head related events

ghost commented:
@ch1bo I have addressed your comments and tried to get closer to the design we discussed this morning. Feel free to adapt that PR to suits your taste, or just drop it if you have something better in mind.

ghost commented:
Just noticed code formatting is still not right, will try to fix it tomorrow

ch1bo committed:
Minor changelog rewording

ch1bo committed:
Slightly simplify end-to-end test

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -251,11 +266,19 @@ data Chain tx m = Chain
       { observedTx :: OnChainTx tx
       , newChainState :: ChainStateType tx
       }
+  | OtherChainEvent {otherChainEvent :: OtherChainEvent}
```
I suggest (and did push a commit) to not keep this as "open", but specific to the one workaround we have: ignoring a head init tx, which in fact should rather be an 'Observation' (but we can't produce a new chain state for ignored events)

ch1bo committed:
Rename and document ignored init chain event

This makes it clear that this is (currently) an exception and not more
"other chain events" are expected.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -526,38 +526,56 @@ fanout ctx st utxo deadlineSlotNo = do
 
+data NoObservation
+  = NoObservation
+  | ObservedInitTx NotAnInit
```
Naming became a bit contrived here: "observed an init tx, which is not an init". I'll suggest a different name.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -287,23 +292,34 @@ chainSyncHandler tracer callback getTimeHandle ctx localChainState =
+    ObservedInitTx (NotAnInit reason) ->
+      traceWith tracer (InvalidInitTx reason)
+    ObservedInitTx (NotAnInitForUs (mismatchReasonObservation -> RawInitObservation{headId, headPTsNames})) ->
+      traceWith tracer (IgnoredInitTx $ SomeHeadObserved{headId = mkHeadId headId, pubKeyHashes = asChainId <$> headPTsNames})
```
I think we do not need to log these events specifically: 
* If the transaction is not an init tx we should not care (like for the other transactions).
* The `NotAnInitForUs` branch here is not reached as these cases are now converted to a `Right` and yield in a `callback` (which does log the event anyways)

ch1bo committed:
Renamed NoObservation constructor and not log invalid init tx

As with other transactions, we do not log a non-observation. In case of
an init tx which is ignored, the chain event callback will yield the log
entry.

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -214,11 +215,13 @@ processNextEvent ::
  where
   NodeState{modifyHeadState} = nodeState
 
+  computeOutcome = Heads.update (Logic.update env ledger)
```
I like this way of composing logic functions!

We could compose multiple of such in a chain and have a final `\st ev -> Error $ InvalidEvent ev st` component (where the error should be named "Unhandled event" or so)

ch1bo reviewed (approved):
Pushed some changes and I think this is a nice increment and good (temporary) workaround to log/inform about ignored init transactions .. until we will further simplify the "callback direction" with #1096 and get rid of the `IgnoredInitTx :: ChainEvent`.
